In [1]:
import numpy as np
import pandas as pd

In [2]:
import tensorflow
import keras

C:\ProgramData\Anaconda3\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


In [3]:
from keras.datasets import mnist

In [4]:
(train_images, train_labels),(test_images, test_labels) = mnist.load_data()
print(train_images.shape)
print(train_labels.shape)
print(test_images.shape)
print(test_labels.shape)

(60000, 28, 28)
(60000,)
(10000, 28, 28)
(10000,)


<h1>Basic MNIST</h1>

In [5]:
from keras import models
from keras import layers

network = keras.Sequential()
network.add(layers.Dense(512, activation='relu', input_shape=(28*28,)))
network.add(layers.Dense(10, activation='softmax'))

In [6]:
network.compile(optimizer='rmsprop',
               loss='categorical_crossentropy',
               metrics=['accuracy'])

In [7]:
train_images=train_images.reshape(60000,28*28)
train_images = train_images.astype('float32')/255

test_images=test_images.reshape(10000,28*28)
test_images = test_images.astype('float32')/255

In [9]:
from keras.utils import to_categorical

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

network.fit(train_images, train_labels, epochs=5, batch_size=128)

Epoch 1/5
60000/60000 [==============================] - 7s 124us/step - loss: 0.2565 - acc: 0.9257
Epoch 2/5
60000/60000 [==============================] - 6s 107us/step - loss: 0.1042 - acc: 0.9692
Epoch 3/5
60000/60000 [==============================] - 7s 109us/step - loss: 0.0690 - acc: 0.9790
Epoch 4/5
60000/60000 [==============================] - 7s 109us/step - loss: 0.0498 - acc: 0.9852
Epoch 5/5
60000/60000 [==============================] - 7s 111us/step - loss: 0.0380 - acc: 0.9882


In [10]:
test_loss, test_acc = network.evaluate(test_images, test_labels)
print("Test accuracy is {0}".format(test_acc))

10000/10000 [==============================] - 1s 68us/step
Test accuracy is 0.9765


<h3>Proper loss parametes </h3>
<p>Binary classification : binary_crossentropy</p>
<p>Multiclass classification : categorical_crossentropy</p>
<p>Regression : mean_squared_error</p>

<h1>IMDB dataset. Binary classification</h1>

In [3]:
from keras.datasets import imdb

In [19]:
(train_data, train_labels),(test_data, test_labels) = imdb.load_data(num_words=10000)

In [44]:
train_data[0]

AttributeError: 'list' object has no attribute 'shape'

In [ ]:
train_labels[:3]

The train_data and test_data is an array of reviews. A review consists of a list of word indices.

In [24]:
def vectorize_sequence(sequences, dim=10000):
    results = np.zeros((len(sequences), dim))
    for i, sequence in enumerate(sequences):
        results[i,sequence] = 1
    return results

In [25]:
x_train = vectorize_sequence(train_data)
x_test = vectorize_sequence(test_data)

In [45]:
x_train.shape

(25000, 10000)

In [38]:
y_train = np.asarray(train_labels).astype('float32')
y_test = np.asarray(test_labels).astype('float32')

In [39]:
y_train.shape

(25000,)

In [34]:
from keras import models, layers
network = keras.Sequential()
network.add(layers.Dense(16, activation='relu',input_shape=(10000,)))
network.add(layers.Dense(16, activation='relu'))
network.add(layers.Dense(1, activation='sigmoid'))  #as we need probability of success from a single node

In [36]:
network.compile(optimizer="rmsprop",
               loss="mse",
               metrics=['accuracy'])

In [40]:
x_val=x_train[:10000]
x_train_partial = x_train[10000:]

y_val=y_train[:10000]
y_train_partial = y_train[10000:]

In [41]:
network.fit(x_train_partial,
           y_train_partial,
           epochs=10,
           batch_size=512,
           validation_data=(x_val, y_val))

Train on 15000 samples, validate on 10000 samples
Epoch 1/10
15000/15000 [==============================] - 4s 239us/step - loss: 0.5084 - acc: 0.7813 - val_loss: 0.3797 - val_acc: 0.8684
Epoch 2/10
15000/15000 [==============================] - 3s 190us/step - loss: 0.3004 - acc: 0.9046 - val_loss: 0.3004 - val_acc: 0.8897
Epoch 3/10
15000/15000 [==============================] - 3s 196us/step - loss: 0.2179 - acc: 0.9285 - val_loss: 0.3086 - val_acc: 0.8710
Epoch 4/10
15000/15000 [==============================] - 3s 212us/step - loss: 0.1751 - acc: 0.9439 - val_loss: 0.2840 - val_acc: 0.8832
Epoch 5/10
15000/15000 [==============================] - 3s 192us/step - loss: 0.1427 - acc: 0.9543 - val_loss: 0.2842 - val_acc: 0.8871
Epoch 6/10
15000/15000 [==============================] - 3s 193us/step - loss: 0.1150 - acc: 0.9651 - val_loss: 0.3166 - val_acc: 0.8771
Epoch 7/10
15000/15000 [==============================] - 3s 194us/step - loss: 0.0980 - acc: 0.9707 - val_loss: 0.3127 - 

In [42]:
evaluate_test_results = network.evaluate(x_test, y_test)

25000/25000 [==============================] - 4s 148us/step


In [43]:
loss, accuracy = evaluate_test_results
#first one is loss and secodn one is accuracy on test set

[0.41986478815078737, 0.86468]

Thus we have attained an accuracy of 88% on the test set. We can use the predict method on x_test to get probability predictions. Whoopee!!